File: crates/rocksky-sdk/src/remote_controller.rs
//! Build a Rocksky remote controller.
//!
//! [`RemoteController`] is the other half of the remote-control WebSocket
//! protocol (see `remote-ws/PROTOCOL.md`): where [`RemotePlayer`](crate::RemotePlayer)
//! is a controllable player, a controller is the remote UI — it lists the user's
//! players, shows what each is playing (now-playing / status / queue), selects
//! the primary device, and sends commands (play / pause / next / previous / seek
//! / queue actions / enqueue). Heartbeat, reconnect, and the device-id handshake
//! are handled for you.
//!
//! **Poll model.** Incoming updates are delivered through [`RemoteController::next_event`]
//! rather than callbacks — the host drives a simple loop. This keeps the API
//! identical across every FFI binding (UniFFI, C-ABI, NIF), none of which need a
//! callback-into-foreign-code path.
//!
//! ```no_run
//! # use rocksky_sdk::remote_controller::{RemoteController, RemoteControllerConfig, RemoteEvent};
//! # async fn ex() {
//! let controller = RemoteController::connect(RemoteControllerConfig::new("<token>", "My Controller"));
//! while let Some(event) = controller.next_event().await {
//!     match event {
//!         RemoteEvent::Devices { devices, .. } => { /* render the picker */ let _ = devices; }
//!         RemoteEvent::NowPlaying { device_id, track, .. } => {
//!             controller.play(Some(device_id)); // e.g. resume that device
//!             let _ = track;
//!         }
//!         _ => {}
//!     }
//! }
//! # }
//! ```
//!
//! Gated behind the `remote-player` feature.

use std::time::Duration;

use futures::{SinkExt, StreamExt};
use serde_json::{json, Value};
use tokio::sync::{mpsc, watch, Mutex};
use tokio_tungstenite::connect_async;
use tokio_tungstenite::tungstenite::Message;

use crate::remote_player::{RemoteNowPlaying, RemoteQueueItem, RemoteStatus, DEFAULT_REMOTE_WS};

const HEARTBEAT: Duration = Duration::from_secs(10);
const RECONNECT: Duration = Duration::from_secs(3);

/// Configuration for [`RemoteController::connect`].
#[derive(Clone, Debug)]
pub struct RemoteControllerConfig {
    /// Rocksky access token (a JWT — from `rocksky login` or an access token).
    pub token: String,
    /// Registration label. Controllers are hidden from device lists, so this is
    /// mostly for the server's logs.
    pub name: String,
    /// WebSocket endpoint. Defaults to [`DEFAULT_REMOTE_WS`].
    pub url: String,
}

impl RemoteControllerConfig {
    /// Config with the default endpoint.
    pub fn new(token: impl Into<String>, name: impl Into<String>) -> Self {
        Self {
            token: token.into(),
            name: name.into(),
            url: DEFAULT_REMOTE_WS.to_string(),
        }
    }

    /// Override the endpoint (builder-style).
    pub fn url(mut self, url: impl Into<String>) -> Self {
        self.url = url.into();
        self
    }
}

/// A player device visible to the controller (from the `devices` snapshot).
#[derive(Clone, Debug, Default)]
pub struct RemoteDevice {
    pub device_id: String,
    pub name: String,
    /// The device's current track, if it has advertised one.
    pub now_playing: Option<RemoteNowPlaying>,
    /// Active index within [`queue`](Self::queue).
    pub queue_index: u32,
    pub queue: Vec<RemoteQueueItem>,
}

/// An update pushed to the controller by the server.
#[derive(Clone, Debug)]
pub enum RemoteEvent {
    /// The full device snapshot (sent right after (re)register).
    Devices {
        primary_device: Option<String>,
        devices: Vec<RemoteDevice>,
    },
    /// A device joined.
    DeviceRegistered { device_id: String, name: String },
    /// A device left.
    DeviceUnregistered { device_id: String },
    /// The primary (scrobble/profile) device changed.
    PrimaryChanged { device_id: String },
    /// A device advertised a new track.
    NowPlaying {
        device_id: String,
        device_name: String,
        track: RemoteNowPlaying,
    },
    /// A device advertised a transport-state change.
    Status {
        device_id: String,
        device_name: String,
        status: RemoteStatus,
    },
    /// A device advertised its queue.
    Queue {
        device_id: String,
        device_name: String,
        index: u32,
        queue: Vec<RemoteQueueItem>,
    },
}

// Outbound frames carried from the handle to the background task, which stamps
// them with the token before sending.
enum CtlOut {
    SetPrimary(String),
    Command {
        action: String,
        target: Option<String>,
        args: Option<Value>,
    },
}

/// A remote controller. Cheap to hold; the connection runs in a background task
/// on the current tokio runtime until this handle is dropped (or
/// [`disconnect`](RemoteController::disconnect) is called).
pub struct RemoteController {
    out_tx: mpsc::UnboundedSender<CtlOut>,
    event_rx: Mutex<mpsc::UnboundedReceiver<RemoteEvent>>,
    stop_tx: watch::Sender<bool>,
}

impl RemoteController {
    /// Connect, register, and maintain the session (heartbeat + auto-reconnect)
    /// in a background task. Must be called from within a tokio runtime.
    pub fn connect(config: RemoteControllerConfig) -> Self {
        let (out_tx, out_rx) = mpsc::unbounded_channel();
        let (event_tx, event_rx) = mpsc::unbounded_channel();
        let (stop_tx, stop_rx) = watch::channel(false);

        tokio::spawn(background(config, out_rx, event_tx, stop_rx));

        Self {
            out_tx,
            event_rx: Mutex::new(event_rx),
            stop_tx,
        }
    }

    /// Await the next update from the server. Returns `None` once the controller
    /// is disconnected (handle dropped or [`disconnect`](Self::disconnect) called).
    pub async fn next_event(&self) -> Option<RemoteEvent> {
        self.event_rx.lock().await.recv().await
    }

    /// Choose the primary (scrobble/profile) device.
    pub fn set_primary(&self, device_id: impl Into<String>) {
        let _ = self.out_tx.send(CtlOut::SetPrimary(device_id.into()));
    }

    /// Send a `play` command. `target` = a device id, or `None` to broadcast to
    /// all the user's devices.
    pub fn play(&self, target: Option<String>) {
        self.command("play", target, None);
    }

    /// Send a `pause` command.
    pub fn pause(&self, target: Option<String>) {
        self.command("pause", target, None);
    }

    /// Send a `next` command.
    pub fn next(&self, target: Option<String>) {
        self.command("next", target, None);
    }

    /// Send a `previous` command.
    pub fn previous(&self, target: Option<String>) {
        self.command("previous", target, None);
    }

    /// Seek the target device to `position_ms`.
    pub fn seek(&self, target: Option<String>, position_ms: u64) {
        self.command("seek", target, Some(json!({ "position": position_ms })));
    }

    /// Jump to `index` in the target device's queue.
    pub fn queue_jump(&self, target: Option<String>, index: u32) {
        self.command("queue_jump", target, Some(json!({ "index": index })));
    }

    /// Remove queue item `index` on the target device.
    pub fn queue_remove(&self, target: Option<String>, index: u32) {
        self.command("queue_remove", target, Some(json!({ "index": index })));
    }

    /// Enqueue tracks on the target device. `mode` is `"now"` | `"next"` |
    /// `"last"`; `start_index` is the entry to start at when `mode == "now"`.
    pub fn enqueue(
        &self,
        target: Option<String>,
        tracks: Vec<RemoteQueueItem>,
        mode: impl Into<String>,
        shuffle: bool,
        start_index: u32,
    ) {
        let args = json!({
            "tracks": tracks.iter().map(descriptor_json).collect::<Vec<_>>(),
            "mode": mode.into(),
            "shuffle": shuffle,
            "startIndex": start_index,
        });
        self.command("enqueue", target, Some(args));
    }

    /// Send an arbitrary command (escape hatch).
    pub fn command(&self, action: impl Into<String>, target: Option<String>, args: Option<Value>) {
        let _ = self.out_tx.send(CtlOut::Command {
            action: action.into(),
            target,
            args,
        });
    }

    /// Stop the connection and end [`next_event`](Self::next_event).
    pub fn disconnect(&self) {
        let _ = self.stop_tx.send(true);
    }
}

impl Drop for RemoteController {
    fn drop(&mut self) {
        let _ = self.stop_tx.send(true);
    }
}

// ── Background task ───────────────────────────────────────────────────────────

async fn background(
    config: RemoteControllerConfig,
    mut out_rx: mpsc::UnboundedReceiver<CtlOut>,
    event_tx: mpsc::UnboundedSender<RemoteEvent>,
    mut stop_rx: watch::Receiver<bool>,
) {
    loop {
        if *stop_rx.borrow() {
            return;
        }

        tokio::select! {
            _ = stop_rx.changed() => return,
            _ = session(&config, &mut out_rx, &event_tx) => {}
        }

        tokio::select! {
            _ = stop_rx.changed() => return,
            _ = tokio::time::sleep(RECONNECT) => {}
        }
    }
}

/// One connect → register → read/write session. Returns when the socket drops
/// (so the caller reconnects). The server re-sends the `devices` snapshot on
/// re-register, so no local resync is needed.
async fn session(
    config: &RemoteControllerConfig,
    out_rx: &mut mpsc::UnboundedReceiver<CtlOut>,
    event_tx: &mpsc::UnboundedSender<RemoteEvent>,
) {
    let (ws, _) = match connect_async(&config.url).await {
        Ok(ok) => ok,
        Err(e) => {
            tracing::warn!(error = %e, "remote-controller connect failed");
            return;
        }
    };
    tracing::info!(url = %config.url, "remote-controller connected");
    let (mut write, mut read) = ws.split();

    // Register (controllers register too; they're hidden from device lists).
    let register = json!({ "type": "register", "clientName": config.name, "token": config.token });
    if write
        .send(Message::Text(register.to_string().into()))
        .await
        .is_err()
    {
        return;
    }

    let mut heartbeat = tokio::time::interval(HEARTBEAT);
    heartbeat.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);

    loop {
        tokio::select! {
            // Outbound command / set_primary from the host.
            msg = out_rx.recv() => {
                let Some(msg) = msg else { return; }; // handle dropped
                if send_out(&mut write, config, &msg).await.is_err() {
                    return;
                }
            }

            // Inbound frame from the server.
            frame = read.next() => {
                match frame {
                    Some(Ok(Message::Text(text))) => {
                        for event in parse_events(text.as_str()) {
                            if event_tx.send(event).is_err() {
                                return; // receiver gone
                            }
                        }
                    }
                    Some(Ok(Message::Close(_))) | None => return,
                    Some(Err(e)) => {
                        tracing::warn!(error = %e, "remote-controller read error");
                        return;
                    }
                    Some(Ok(_)) => {} // ping/pong/binary
                }
            }

            // Heartbeat.
            _ = heartbeat.tick() => {
                if write.send(Message::Text("ping".to_string().into())).await.is_err() {
                    return;
                }
            }
        }
    }
}

async fn send_out<S>(write: &mut S, config: &RemoteControllerConfig, msg: &CtlOut) -> Result<(), ()>
where
    S: SinkExt<Message> + Unpin,
{
    let frame = match msg {
        CtlOut::SetPrimary(id) => json!({
            "type": "set_primary",
            "device_id": id,
            "token": config.token,
        }),
        CtlOut::Command {
            action,
            target,
            args,
        } => {
            let mut o = json!({ "type": "command", "action": action, "token": config.token });
            if let Some(t) = target {
                o["target"] = json!(t);
            }
            if let Some(a) = args {
                o["args"] = a.clone();
            }
            o
        }
    };
    write
        .send(Message::Text(frame.to_string().into()))
        .await
        .map_err(|_| ())
}

// ── Inbound parsing ───────────────────────────────────────────────────────────

/// Parse a server frame into zero or more [`RemoteEvent`]s.
fn parse_events(text: &str) -> Vec<RemoteEvent> {
    if text == "pong" {
        return Vec::new();
    }
    let Ok(v) = serde_json::from_str::<Value>(text) else {
        return Vec::new();
    };
    // The registration reply carries no event (we don't need our own id to
    // control other devices).
    if v.get("status").and_then(|x| x.as_str()) == Some("registered") {
        return Vec::new();
    }
    let ty = v.get("type").and_then(|x| x.as_str()).unwrap_or("");
    match ty {
        "devices" => vec![RemoteEvent::Devices {
            primary_device: v
                .get("primary_device")
                .and_then(|x| x.as_str())
                .map(str::to_string),
            devices: v
                .get("devices")
                .and_then(|x| x.as_array())
                .map(|a| a.iter().map(device_from_value).collect())
                .unwrap_or_default(),
        }],
        "device_registered" => vec![RemoteEvent::DeviceRegistered {
            device_id: str_field(&v, "deviceId"),
            name: str_field(&v, "clientName"),
        }],
        "device_unregistered" => vec![RemoteEvent::DeviceUnregistered {
            device_id: str_field(&v, "device_id"),
        }],
        "primary_changed" => vec![RemoteEvent::PrimaryChanged {
            device_id: str_field(&v, "device_id"),
        }],
        "message" => message_event(&v).into_iter().collect(),
        _ => Vec::new(),
    }
}

/// A `message` broadcast wraps a device's `track` / `status` / `queue` state.
fn message_event(v: &Value) -> Option<RemoteEvent> {
    let device_id = str_field(v, "device_id");
    let device_name = str_field(v, "device_name");
    let data = v.get("data")?;
    match data.get("type").and_then(|x| x.as_str())? {
        "track" => Some(RemoteEvent::NowPlaying {
            device_id,
            device_name,
            track: track_from_value(data),
        }),
        "status" => Some(RemoteEvent::Status {
            device_id,
            device_name,
            status: status_from_code(data.get("status").and_then(|x| x.as_i64()).unwrap_or(0)),
        }),
        "queue" => Some(RemoteEvent::Queue {
            device_id,
            device_name,
            index: data.get("index").and_then(|x| x.as_u64()).unwrap_or(0) as u32,
            queue: data
                .get("queue")
                .and_then(|x| x.as_array())
                .map(|a| a.iter().map(queue_item_from_value).collect())
                .unwrap_or_default(),
        }),
        _ => None,
    }
}

fn device_from_value(v: &Value) -> RemoteDevice {
    let now_playing = v
        .get("now_playing")
        .filter(|x| x.is_object())
        .map(track_from_value);
    let (queue_index, queue) = v
        .get("queue")
        .map(|q| {
            (
                q.get("index").and_then(|x| x.as_u64()).unwrap_or(0) as u32,
                q.get("queue")
                    .and_then(|x| x.as_array())
                    .map(|a| a.iter().map(queue_item_from_value).collect())
                    .unwrap_or_default(),
            )
        })
        .unwrap_or((0, Vec::new()));
    RemoteDevice {
        device_id: str_field(v, "device_id"),
        name: str_field(v, "name"),
        now_playing,
        queue_index,
        queue,
    }
}

fn track_from_value(v: &Value) -> RemoteNowPlaying {
    RemoteNowPlaying {
        title: str_field(v, "title"),
        artist: str_field(v, "artist"),
        album: str_field(v, "album"),
        album_artist: str_field(v, "album_artist"),
        album_art: str_field(v, "album_art"),
        duration_ms: v
            .get("duration_ms")
            .and_then(|x| x.as_u64())
            .or_else(|| v.get("length").and_then(|x| x.as_u64()))
            .unwrap_or(0),
        elapsed_ms: v.get("elapsed").and_then(|x| x.as_u64()).unwrap_or(0),
        is_playing: v
            .get("is_playing")
            .and_then(|x| x.as_bool())
            .unwrap_or(false),
        codec: v.get("codec").and_then(|x| x.as_str()).map(str::to_string),
        sample_rate: v
            .get("sample_rate")
            .and_then(|x| x.as_u64())
            .map(|n| n as u32),
    }
}

fn queue_item_from_value(v: &Value) -> RemoteQueueItem {
    RemoteQueueItem {
        upload_id: str_field(v, "uploadId"),
        track_id: str_field(v, "trackId"),
        title: str_field(v, "title"),
        artist: str_field(v, "artist"),
        album: str_field(v, "album"),
        album_artist: str_field(v, "album_artist"),
        album_art: str_field(v, "album_art"),
        duration_ms: v.get("duration").and_then(|x| x.as_u64()).unwrap_or(0),
        song_uri: str_field(v, "song_uri"),
        album_uri: str_field(v, "album_uri"),
        track_number: v.get("track_number").and_then(|x| x.as_i64()).unwrap_or(0) as i32,
    }
}

/// An `enqueue` descriptor the controller sends to a player.
fn descriptor_json(t: &RemoteQueueItem) -> Value {
    json!({
        "uploadId": t.upload_id,
        "trackId": t.track_id,
        "title": t.title,
        "artist": t.artist,
        "album": t.album,
        "album_artist": t.album_artist,
        "album_art": t.album_art,
        "duration": t.duration_ms,
        "song_uri": t.song_uri,
        "album_uri": t.album_uri,
        "track_number": t.track_number,
    })
}

fn status_from_code(code: i64) -> RemoteStatus {
    match code {
        1 => RemoteStatus::Playing,
        0 => RemoteStatus::Stopped,
        _ => RemoteStatus::Paused, // 2 or 3
    }
}

fn str_field(v: &Value, key: &str) -> String {
    v.get(key)
        .and_then(|x| x.as_str())
        .unwrap_or("")
        .to_string()
}

File: sdk/typescript/src/remote-controller.ts
/**
 * RemoteController — build a Rocksky remote UI in a few lines.
 *
 * The other half of the remote-control WebSocket protocol (see
 * remote-ws/PROTOCOL.md): where {@link RemotePlayer} is a controllable player, a
 * controller lists the user's players, observes what each is playing
 * (now-playing / status / queue), picks the primary device, and sends commands
 * (play/pause/next/previous/seek/queue actions/enqueue). Heartbeat, reconnect,
 * and the register handshake are handled for you.
 *
 * ```ts
 * const controller = new RemoteController({ token, name: "My Controller" });
 * controller.on("devices", (d) => render(d.devices));
 * controller.on("nowPlaying", (e) => showTrack(e.deviceId, e.track));
 * controller.connect();
 * // …then drive a device:
 * controller.setPrimary(deviceId);
 * controller.pause(deviceId);
 * controller.seek(deviceId, 42000);
 * ```
 */

import {
  DEFAULT_REMOTE_WS,
  type RemoteAudioSettings,
  type RemoteNowPlaying,
  type RemoteQueueItem,
  type RemoteRepeat,
} from "./remote-player.js";

export { DEFAULT_REMOTE_WS, type RemoteNowPlaying, type RemoteQueueItem };

/** A player device visible to the controller. */
export interface RemoteDevice {
  deviceId: string;
  name: string;
  nowPlaying?: RemoteNowPlaying;
  queueIndex: number;
  queue: RemoteQueueItem[];
}

/** Transport state advertised by a device. */
export type RemoteStatus = "playing" | "paused" | "stopped";

/** Events the controller emits (from the server). */
export interface RemoteControllerEvents {
  /** The full device snapshot (on connect + reconnect). */
  devices(payload: { primaryDevice: string | null; devices: RemoteDevice[] }): void;
  deviceRegistered(payload: { deviceId: string; name: string }): void;
  deviceUnregistered(payload: { deviceId: string }): void;
  primaryChanged(payload: { deviceId: string }): void;
  nowPlaying(payload: { deviceId: string; deviceName: string; track: RemoteNowPlaying }): void;
  status(payload: { deviceId: string; deviceName: string; status: RemoteStatus }): void;
  queue(payload: {
    deviceId: string;
    deviceName: string;
    index: number;
    queue: RemoteQueueItem[];
  }): void;
}

type EventName = keyof RemoteControllerEvents;
type EventHandler<K extends EventName> = RemoteControllerEvents[K];

export interface RemoteControllerOptions {
  /** Rocksky access token, or a getter (read fresh on each (re)connect/send). */
  token: string | (() => string | undefined);
  /** Registration label (controllers are hidden from device lists). */
  name: string;
  /** WebSocket endpoint. Defaults to {@link DEFAULT_REMOTE_WS}. */
  url?: string;
  /** Heartbeat interval, ms (default 10000). */
  heartbeatMs?: number;
  /** Reconnect delay after a drop, ms (default 3000). */
  reconnectMs?: number;
  /** Optional debug logger. */
  debug?: (...args: unknown[]) => void;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Json = any;

export class RemoteController {
  private ws: WebSocket | null = null;
  private stopped = false;
  private heartbeat?: ReturnType<typeof setInterval>;
  private reconnectTimer?: ReturnType<typeof setTimeout>;

  private readonly url: string;
  private readonly heartbeatMs: number;
  private readonly reconnectMs: number;
  private readonly getToken: () => string | undefined;
  private readonly debug: (...args: unknown[]) => void;

  private handlers: { [K in EventName]?: EventHandler<K> } = {};

  constructor(private readonly opts: RemoteControllerOptions) {
    this.url = opts.url ?? DEFAULT_REMOTE_WS;
    this.heartbeatMs = opts.heartbeatMs ?? 10_000;
    this.reconnectMs = opts.reconnectMs ?? 3_000;
    this.getToken =
      typeof opts.token === "function" ? opts.token : () => opts.token as string;
    this.debug = opts.debug ?? (() => {});
  }

  /** Subscribe to a server event. Returns `this` for chaining. */
  on<K extends EventName>(event: K, handler: EventHandler<K>): this {
    this.handlers[event] = handler;
    return this;
  }

  /** Connect, register, and start the heartbeat + auto-reconnect loop. */
  connect(): void {
    this.stopped = false;
    this.open();
  }

  /** Tear the connection down and stop reconnecting. */
  disconnect(): void {
    this.stopped = true;
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    if (this.heartbeat) clearInterval(this.heartbeat);
    try {
      this.ws?.close();
    } catch {
      /* ignore */
    }
    this.ws = null;
  }

  // ── Controls ──────────────────────────────────────────────────────────────

  /** Choose the primary (scrobble/profile) device. */
  setPrimary(deviceId: string): void {
    this.send({ type: "set_primary", device_id: deviceId, token: this.getToken() });
  }

  /** `play`. `target` is a device id, or omit to broadcast to all devices. */
  play(target?: string): void {
    this.command("play", target);
  }

  pause(target?: string): void {
    this.command("pause", target);
  }

  next(target?: string): void {
    this.command("next", target);
  }

  previous(target?: string): void {
    this.command("previous", target);
  }

  seek(target: string | undefined, positionMs: number): void {
    this.command("seek", target, { position: positionMs });
  }

  queueJump(target: string | undefined, index: number): void {
    this.command("queue_jump", target, { index });
  }

  queueRemove(target: string | undefined, index: number): void {
    this.command("queue_remove", target, { index });
  }

  /** Enqueue tracks on the target device. */
  enqueue(
    target: string | undefined,
    tracks: RemoteQueueItem[],
    mode: "now" | "next" | "last" = "now",
    shuffle = false,
    startIndex = 0,
  ): void {
    this.command("enqueue", target, {
      tracks: tracks.map((t) => ({
        uploadId: t.uploadId,
        trackId: t.trackId,
        title: t.title,
        artist: t.artist,
        album: t.album,
        album_artist: t.albumArtist,
        album_art: t.albumArt,
        duration: t.durationMs,
        song_uri: t.songUri,
        album_uri: t.albumUri,
        track_number: t.trackNumber,
      })),
      mode,
      shuffle,
      startIndex,
    });
  }

  /** Turn queue shuffle on/off. A player without shuffle ignores it. */
  setShuffle(target: string | undefined, enabled: boolean): void {
    this.command("shuffle", target, { enabled });
  }

  /** Set the queue repeat mode. A player without repeat ignores it. */
  setRepeat(target: string | undefined, mode: RemoteRepeat): void {
    this.command("repeat", target, { mode });
  }

  /** Set output volume, 0–1. A player without volume control ignores it. */
  setVolume(target: string | undefined, volume: number): void {
    this.command("volume", target, {
      volume: Math.min(1, Math.max(0, volume)),
    });
  }

  /**
   * Apply a partial audio-settings document. Safe to send verbatim to any
   * device: every section is optional and a player applies only the ones its
   * engine implements, so there is nothing to probe or negotiate first.
   */
  setAudioSettings(
    target: string | undefined,
    settings: RemoteAudioSettings,
  ): void {
    this.command("audio_settings", target, settings as unknown as Json);
  }

  /** Send an arbitrary command (escape hatch). */
  command(action: string, target?: string, args?: Json): void {
    const payload: Json = { type: "command", action, token: this.getToken() };
    if (target) payload.target = target;
    if (args !== undefined) payload.args = args;
    this.send(payload);
  }

  // ── Internals ───────────────────────────────────────────────────────────────

  private send(payload: Json): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      try {
        this.ws.send(JSON.stringify(payload));
      } catch (e) {
        this.debug("send error", e);
      }
    }
  }

  private open(): void {
    if (this.stopped) return;
    const token = this.getToken();
    if (!token) {
      this.reconnectTimer = setTimeout(() => this.open(), this.reconnectMs);
      return;
    }

    let ws: WebSocket;
    try {
      ws = new WebSocket(this.url);
    } catch (e) {
      this.debug("connect failed", e);
      this.reconnectTimer = setTimeout(() => this.open(), this.reconnectMs);
      return;
    }
    this.ws = ws;

    ws.onopen = () => {
      this.debug("connected");
      this.send({ type: "register", clientName: this.opts.name, token: this.getToken() });
      if (this.heartbeat) clearInterval(this.heartbeat);
      this.heartbeat = setInterval(() => {
        if (ws.readyState === WebSocket.OPEN) ws.send("ping");
      }, this.heartbeatMs);
    };

    ws.onmessage = (ev: MessageEvent) => {
      if (ev.data === "pong") return;
      let msg: Json;
      try {
        msg = JSON.parse(ev.data as string);
      } catch {
        return;
      }
      this.handle(msg);
    };

    ws.onerror = () => {
      try {
        ws.close();
      } catch {
        /* ignore */
      }
    };

    ws.onclose = () => {
      this.debug("disconnected");
      if (this.heartbeat) clearInterval(this.heartbeat);
      if (this.ws === ws) this.ws = null;
      if (!this.stopped) {
        if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
        this.reconnectTimer = setTimeout(() => this.open(), this.reconnectMs);
      }
    };
  }

  private handle(msg: Json): void {
    // The registration reply carries no event (a controller doesn't need its own
    // id to control other devices).
    if (msg.status === "registered") return;
    switch (msg.type) {
      case "devices":
        this.handlers.devices?.({
          primaryDevice: msg.primary_device ?? null,
          devices: (msg.devices ?? []).map(deviceFromJson),
        });
        break;
      case "device_registered":
        this.handlers.deviceRegistered?.({
          deviceId: msg.deviceId ?? "",
          name: msg.clientName ?? "",
        });
        break;
      case "device_unregistered":
        this.handlers.deviceUnregistered?.({ deviceId: msg.device_id ?? "" });
        break;
      case "primary_changed":
        this.handlers.primaryChanged?.({ deviceId: msg.device_id ?? "" });
        break;
      case "message":
        this.handleMessage(msg);
        break;
      default:
        this.debug("unknown frame", msg.type);
    }
  }

  private handleMessage(msg: Json): void {
    const deviceId = msg.device_id ?? "";
    const deviceName = msg.device_name ?? "";
    const data = msg.data;
    if (!data) return;
    switch (data.type) {
      case "track":
        this.handlers.nowPlaying?.({ deviceId, deviceName, track: trackFromJson(data) });
        break;
      case "status":
        this.handlers.status?.({ deviceId, deviceName, status: statusFromCode(data.status) });
        break;
      case "queue":
        this.handlers.queue?.({
          deviceId,
          deviceName,
          index: data.index ?? 0,
          queue: (data.queue ?? []).map(queueItemFromJson),
        });
        break;
    }
  }
}

function statusFromCode(code: number): RemoteStatus {
  return code === 1 ? "playing" : code === 0 ? "stopped" : "paused";
}

function trackFromJson(d: Json): RemoteNowPlaying {
  return {
    title: d.title ?? "",
    artist: d.artist ?? "",
    album: d.album,
    albumArtist: d.album_artist,
    albumArt: d.album_art,
    durationMs: d.duration_ms ?? d.length,
    elapsedMs: d.elapsed,
    isPlaying: d.is_playing,
    codec: d.codec,
    sampleRate: d.sample_rate,
    // Server-enriched fields (present on the broadcast a controller receives).
    songUri: d.song_uri,
    albumUri: d.album_uri,
    artistUri: d.artist_uri,
    sha256: d.sha256,
    liked: d.liked,
  };
}

function queueItemFromJson(d: Json): RemoteQueueItem {
  return {
    uploadId: d.uploadId,
    trackId: d.trackId,
    title: d.title ?? "",
    artist: d.artist ?? "",
    album: d.album,
    albumArtist: d.album_artist,
    albumArt: d.album_art,
    durationMs: d.duration,
    songUri: d.song_uri,
    albumUri: d.album_uri,
    trackNumber: d.track_number,
  };
}

function deviceFromJson(d: Json): RemoteDevice {
  return {
    deviceId: d.device_id ?? "",
    name: d.name ?? "",
    nowPlaying: d.now_playing ? trackFromJson(d.now_playing) : undefined,
    queueIndex: d.queue?.index ?? 0,
    queue: (d.queue?.queue ?? []).map(queueItemFromJson),
  };
}
